File: public/rust/crates/ddk-rs/examples/ddk_toy/src/binding.c
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <ddk/device.h>
#include <ddk/driver.h>
#include <ddk/binding.h>

extern zx_driver_ops_t DRIVER_OPS_DDK_TOY;

ZIRCON_DRIVER_BEGIN(_driver_ddk_toy, DRIVER_OPS_DDK_TOY, "ddk-toy", "0.1.0", 1)
    BI_MATCH_IF(EQ, BIND_PROTOCOL, ZX_PROTOCOL_MISC_PARENT),
ZIRCON_DRIVER_END(_driver_ddk_toy)

File: src/tests/benchmarks/fidl/roundtrip/runner/src/binding.rs
// Copyright 2020 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

use {
    fidl_fidl_benchmarks::BindingsUnderTestMarker, fidl_fidl_benchmarks::BindingsUnderTestProxy,
    fidl_fuchsia_sys::LauncherProxy, fuchsia_component::client::launch,
    fuchsia_component::client::App, std::sync::Arc,
};

pub struct BindingConfig {
    pub name: &'static str,
    pub url: &'static str,
}

impl BindingConfig {
    pub fn launch(&self, launcher: &LauncherProxy) -> LaunchedBinding {
        println!("Launching {} from {}", self.name, self.url);
        let app = launch(&launcher, self.url.to_string(), None).unwrap();
        let proxy = app.connect_to_service::<BindingsUnderTestMarker>().unwrap();
        LaunchedBinding {
            name: self.name,
            url: self.url,
            proxy: Arc::new(proxy),
            app: Arc::new(app),
        }
    }
}

#[derive(Clone)]
pub struct LaunchedBinding {
    pub name: &'static str,
    pub url: &'static str,
    pub proxy: Arc<BindingsUnderTestProxy>,
    pub app: Arc<App>,
}
